include/libcyphal/transport/can/delegate.hpp

```diff
@@ -0,0 +1,215 @@
+        {
+            return payload_size_;
+        }
+
+        CETL_NODISCARD std::size_t copy(const std::size_t offset_bytes,
+                                        void* const       destination,
+                                        const std::size_t length_bytes) const override
+        {
+            CETL_DEBUG_ASSERT((destination != nullptr) || (length_bytes == 0),
+                              "Destination could be null only with zero bytes ask.");
+
+            if ((destination == nullptr) || (buffer_ == nullptr) || (payload_size_ <= offset_bytes))
+            {
+                return 0;
+            }
+
+            const auto bytes_to_copy = std::min(length_bytes, payload_size_ - offset_bytes);
+            memcpy(destination, static_cast<cetl::byte*>(buffer_) + offset_bytes, bytes_to_copy);
+            return bytes_to_copy;
+        }
+
+    private:
+        // MARK: Data members:
+
+        TransportDelegate& delegate_;
+        void*              buffer_;
+        std::size_t        payload_size_;
+
+    };  // CanardMemory
+
+    explicit TransportDelegate(cetl::pmr::memory_resource& memory)
+        : memory_{memory}
+        , canard_instance_(canardInit(canardMemoryAllocate, canardMemoryFree))
+    {
+        canard_instance().user_reference = this;
+    }
+
+    CETL_NODISCARD inline CanardInstance& canard_instance() noexcept
+    {
+        return canard_instance_;
+    }
+
+    CETL_NODISCARD inline const CanardInstance& canard_instance() const noexcept
+    {
+        return canard_instance_;
+    }
+
+    CETL_NODISCARD inline cetl::pmr::memory_resource& memory() const noexcept
+    {
+        return memory_;
+    }
+
+    static cetl::optional<AnyError> anyErrorFromCanard(const int8_t result)
+    {
+        // Canard error results are negative, so we need to negate them to get the error code.
+        const auto canard_error = static_cast<int8_t>(-result);
+
+        if (canard_error == CANARD_ERROR_INVALID_ARGUMENT)
+        {
+            return ArgumentError{};
+        }
+        if (canard_error == CANARD_ERROR_OUT_OF_MEMORY)
+        {
+            return MemoryError{};
+        }
+
+        return {};
+    }
+
+    void canardMemoryFree(void* const pointer)
+    {
+        if (pointer == nullptr)
+        {
+            return;
+        }
+
+        auto memory_header = static_cast<CanardMemoryHeader*>(pointer);
+        --memory_header;
+
+        memory_.deallocate(memory_header, memory_header->size);
+    }
+
+private:
+    // Until "canardMemFree must provide size" issue #216 is resolved,
+    // we need to store the size of the memory allocated.
+    // TODO: Remove this workaround when the issue is resolved.
+    // see https://github.com/OpenCyphal/libcanard/issues/216
+    //
+    struct CanardMemoryHeader final
+    {
+        alignas(std::max_align_t) std::size_t size;
+    };
+
+    CETL_NODISCARD static inline TransportDelegate& getSelfFrom(const CanardInstance* const ins)
```

Comment:
This is used to ensure compliance with AUTOSAR A0-1-2

Reply:
Compilers can use `-Wunused-result` for this. Furthermore, neither the C++ standard library nor AUTOSAR dictate noexcept on every method. We should only provide this on an expression that might otherwise be considered a discarded-value expression by the compiler.